File: app/src/main/java/com/example/pam_app/adapter/BucketEntryAdapter.java
package com.example.pam_app.adapter;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.pam_app.databinding.ActivityBucketEntryBinding;
import com.example.pam_app.model.Entry;

import java.util.ArrayList;
import java.util.List;

public class BucketEntryAdapter<T extends Entry> extends RecyclerView.Adapter<BucketEntryAdapter.BucketEntryHolder>
        implements BindableAdapter<List<T>> {

    private List<T> items;

    @Override
    public void setData(final List<T> items) {
        this.items = items;
        notifyDataSetChanged();
    }

    public void showNewBucket(final T entry) {
        if (items == null) {
            items = new ArrayList<>();
        }
        items.add(entry);
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public BucketEntryHolder onCreateViewHolder(final ViewGroup parent, final int viewType) {
        final LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        return new BucketEntryHolder(ActivityBucketEntryBinding.inflate(inflater, parent, false));
    }

    @Override
    public int getItemCount() {
        return (items == null) ? 0 : items.size();
    }

    @Override
    public void onBindViewHolder(final BucketEntryHolder holder, final int position) {
        holder.bind(items.get(position));
    }

    public static class BucketEntryHolder extends RecyclerView.ViewHolder {

        private final ActivityBucketEntryBinding binding;

        public BucketEntryHolder(final ActivityBucketEntryBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        public void bind(final Entry entry) {
            this.binding.setEntry(entry);
        }
    }
}

File: app/src/main/java/com/example/pam_app/adapter/BucketEntryAdapter.kt
package com.example.pam_app.adapter

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.example.pam_app.adapter.BucketEntryAdapter.BucketEntryHolder
import com.example.pam_app.databinding.ActivityBucketEntryBinding
import com.example.pam_app.model.Entry
import java.util.*

class BucketEntryAdapter<T : Entry?> : RecyclerView.Adapter<BucketEntryHolder>(), BindableAdapter<Collection<T>?> {
    private var items: LinkedList<T>? = null

    override fun setData(items: Collection<T>?) {
        this.items = LinkedList()
        this.items?.addAll(items!!)
        notifyDataSetChanged()
    }

    fun showNewBucket(entry: T) {
        if (items == null) {
            items = LinkedList()
        }
        items?.add(entry)
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BucketEntryHolder {
        val inflater = LayoutInflater.from(parent.context)
        return BucketEntryHolder(ActivityBucketEntryBinding.inflate(inflater, parent, false))
    }

    override fun getItemCount(): Int {
        return items?.size ?: 0
    }

    override fun onBindViewHolder(holder: BucketEntryHolder, position: Int) {
        holder.bind(items!![position])
    }

    class BucketEntryHolder(private val binding: ActivityBucketEntryBinding) : RecyclerView.ViewHolder(binding.root) {
        fun bind(entry: Entry?) {
            binding.entry = entry
        }

    }
}